### ISSUE TITLE
Can't promote a function returning String with a type signature

### ISSUE THREAD

**Issue by boothead** [goldfirere/singletons#267]
If I have the following:

```
   $(singletons [d|
       data EventScans
        = PassThrough
        | SuiteProgress

       -- cn :: EventScans -> String
       cn PassThrough = "all-events"
       cn SuiteProgress = "suite-progress"
    |])
```
I can't un-comment the type signature for cn.

If I try and use the specific type signature above I get:

```
* Expected kind `[Char]', but `"all-events"' has kind `Symbol'
* In the type `"all-events"'
  In the type family declaration for `Cn' (haskell-stack-ghc)
```
If I leave the signature off I get `cn :: IsString t => EventScans  -> t` as the inferred type.

If I try to use the inferred type in the TH block I get:

```
Variable `a_agPb' used as both a kind and a type
 Did you intend to use TypeInType? (haskell-stack-ghc)
Not in scope: type constructor or class `SIsString'
  Perhaps you meant `IsString' (imported from Data.String) (haskell-stack-ghc)
```
I've tried this with lts-9.2 (and singletons-2.1) and I've also updated to singletons-2.2

There's a stack overflow question here: https://stackoverflow.com/questions/46834285/how-do-i-promote-a-function-that-returns-a-string-with-singletons I'm opening this issue in case this is a bug, it looks a bit similar to #247

**Comment by RyanGlScott** [comment]
OK... and what happens is you enable `TypeInType`? :)

**Comment by boothead** [comment]
I get an error on this class in a different part of the code:

```
class HasProcessor (api :: *) processor where
  type ProcessorT api processor (m :: * -> *) :: *
  server :: Proxy api -> Proxy processor -> Processor api processor

--  Operator applied to too few arguments: *
```

**Comment by RyanGlScott** [comment]
Well, yes. That's to be expected—with `TypeInType` enabled, you have to explicitly import `*` from `Data.Kind`.

Let me put this another way: can you post a complete code example and highlight what you think the issue with `singletons` is? I can't figure out what you think the bug is.

**Comment by boothead** [comment]
The concrete thing I'm asking about is why I can't specify the type signature of the cn function above. If I leave the type signature of the function the code compiles. If I enable `TypeInType` and add the type signature I get the following error:

```
Not in scope: type constructor or class `SIsString'
```

So my question is, is this a bug or user error with promoting a function? Should I expect to be able to include a type signature here, or just leave the type signature off?

**Comment by RyanGlScott** [comment]
OK. So you're trying to give `foo` a type signature in this code:

```haskell
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}

import Data.Singletons.TH

$(singletons
  [d| -- foo :: ???
      foo = "bar"
    |])
```

Such that the type signature works at both the value and singleton level?

If so, `singletons` doesn't support this out-of-the-box at the moment. As you've observed in https://github.com/goldfirere/singletons/issues/267#issue-267084107 (and which I overlooked the first time due to unrelated language extension issues), the problem is that at the value level, `"bar"` will be of type `String`/`Text`/etc., but at the type level, the type-level `"bar"` will be of kind `Symbol`. You'd _like_ to able to write:

```haskell
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}

import Data.Singletons.TH
import Data.String

$(singletons
  [d| foo :: IsString s => s
      foo = "bar"
    |])
```

But at the moment, `singletons` doesn't promote/single the `IsString` class. The good news is that this [will be added in the next version of `singletons`](https://github.com/goldfirere/singletons/pull/268). But if you want this to work today, you can apply these changes to promote/single `IsString` yourself:

```haskell
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE UndecidableInstances #-}

import Data.Singletons.TH
import Data.Singletons.TypeLits
import Data.String

$(singletonsOnly
  [d| class IsString a where
        fromString :: Symbol -> a
    |])

instance PIsString Symbol where
  type FromString a = a

instance SIsString Symbol where
  sFromString x = x

$(singletons
  [d| foo :: IsString s => s
      foo = fromString "bar"
    |])
```

(And after #268 lands, you'll be able to turn on `OverloadedStrings` and just write `foo = "bar"`. But until then, you'll need to write `fromString` explicitly.)

Does that answer your question?

**Comment by boothead** [comment]
Yes, thank you for the very detailed answer! If you don't mind I'll add this to the related stack overflow question?

**Comment by RyanGlScott** [comment]
> If you don't mind I'll add this to the related stack overflow question?

Sure, no objections here.
